### PR TITLE
add code for check_not_valid_constraints.sql

### PR DESCRIPTION
**Pull request by BLoHny**
Relates to https://github.com/mfvanek/pg-index-health/issues/362

### CODE REVIEW

**Comment by BLoHny** [comment]
The current test code doesn't seem to be good code.
Could you help me?

File: pg-index-health/src/main/java/io/github/mfvanek/pg/common/maintenance/Diagnostic.java [hunk 1]
```diff
@@ -38,7 +38,8 @@ public enum Diagnostic {
     COLUMNS_WITHOUT_DESCRIPTION(ExecutionTopology.ON_PRIMARY, "columns_without_description.sql", QueryExecutors::executeQueryWithSchema),
     COLUMNS_WITH_JSON_TYPE(ExecutionTopology.ON_PRIMARY, "columns_with_json_type.sql", QueryExecutors::executeQueryWithSchema),
     COLUMNS_WITH_SERIAL_TYPES(ExecutionTopology.ON_PRIMARY, "non_primary_key_columns_with_serial_types.sql", QueryExecutors::executeQueryWithSchema),
-    FUNCTIONS_WITHOUT_DESCRIPTION(ExecutionTopology.ON_PRIMARY, "functions_without_description.sql", QueryExecutors::executeQueryWithSchema);
+    FUNCTIONS_WITHOUT_DESCRIPTION(ExecutionTopology.ON_PRIMARY, "functions_without_description.sql", QueryExecutors::executeQueryWithSchema),
+    CHECK_NOT_VALID_CONSTRAINTS(ExecutionTopology.ON_PRIMARY, "check_not_valid_constraints.sql", QueryExecutors::executeQueryWithSchema);
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/common/maintenance/Diagnostic.java:42]: I'd prefer just `NOT_VALID_CONSTRAINTS` here

File: pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHost.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import io.github.mfvanek.pg.connection.PgConnection;
+import io.github.mfvanek.pg.model.PgContext;
+import io.github.mfvanek.pg.model.column.Column;
+import io.github.mfvanek.pg.model.constraint.ForeignKey;
+import io.github.mfvanek.pg.utils.ColumnsInForeignKeyParser;
+
+import java.util.List;
+import javax.annotation.Nonnull;
+
+public class CheckNotValidConstraintsOnHost extends AbstractCheckOnHost<ForeignKey> {
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHost.java:13]: Let's rename it to `NotValidConstraintsCheckOnHost`

File: pg-index-health/src/main/java/io/github/mfvanek/pg/checks/cluster/CheckNotValidConstraintsOnCluster.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package io.github.mfvanek.pg.checks.cluster;
+
+import io.github.mfvanek.pg.checks.host.CheckNotValidConstraintsOnHost;
+import io.github.mfvanek.pg.connection.HighAvailabilityPgConnection;
+import io.github.mfvanek.pg.model.constraint.ForeignKey;
+
+import javax.annotation.Nonnull;
+
+public class CheckNotValidConstraintsOnCluster extends AbstractCheckOnCluster<ForeignKey> {
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/checks/cluster/CheckNotValidConstraintsOnCluster.java:9]: Let's rename it to `NotValidConstraintsCheckOnCluster`

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/CheckNotValidConstraintsOnClusterTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import io.github.mfvanek.pg.support.DatabaseAwareTestBase;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.function.Predicate;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CheckNotValidConstraintsOnClusterTest extends DatabaseAwareTestBase {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/CheckNotValidConstraintsOnClusterTest.java:20]: Please rename test class the same as corresponding check-class

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHostTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import io.github.mfvanek.pg.model.column.Column;
+import io.github.mfvanek.pg.model.constraint.ForeignKey;
+import io.github.mfvanek.pg.support.DatabaseAwareTestBase;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static io.github.mfvanek.pg.support.AbstractCheckOnHostAssert.assertThat;
+
+public class CheckNotValidConstraintsOnHostTest extends DatabaseAwareTestBase {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHostTest.java:15]: Please rename test class the same as corresponding check-class

**Review by mfvanek — CHANGES_REQUESTED** [review]
There are checkstyle issues (see tests results).  
Please ensure that project builds locally: `./gradlew build`  
See https://github.com/mfvanek/pg-index-health/blob/master/CONTRIBUTING.md

**Comment by mfvanek** [comment]
> The current test code doesn't seem to be good code. Could you help me?

1. You can copy and modify existing class https://github.com/mfvanek/pg-index-health/blob/master/pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AddLinksBetweenAccountsAndClientsStatement.java in order to bring not valid FK.
2. You can add one more column to https://github.com/mfvanek/pg-index-health/blob/master/pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateTableWithCheckConstraintOnSerialPrimaryKey.java and not valid check constraint on it

File: pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHost.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    public CheckNotValidConstraintsOnHost(@Nonnull final PgConnection pgConnection) {
+        super(ForeignKey.class, pgConnection, Diagnostic.CHECK_NOT_VALID_CONSTRAINTS);
```
Comment (mfvanek) [inline pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHost.java:16]: Please take a look at my comment https://github.com/mfvanek/pg-index-health/issues/362#issuecomment-2044247293

I think we need a new `Constraint `class in model

Comment (mfvanek) [reply pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHost.java:16]:
```java
package io.github.mfvanek.pg.model.constraint;

import javax.annotation.Nonnull;
import java.util.Objects;

public enum ConstraintType {

    CHECK("c"),
    FOREIGN_KEY("f");

    // See https://www.postgresql.org/docs/current/catalog-pg-constraint.html
    private final String pgConType;

    ConstraintType(@Nonnull final String pgConType) {
        this.pgConType = Objects.requireNonNull(pgConType, "pgConType");
    }
}
```

and
```java
package io.github.mfvanek.pg.model.constraint;

import io.github.mfvanek.pg.model.DbObject;
import io.github.mfvanek.pg.model.table.TableNameAware;

import javax.annotation.concurrent.Immutable;

@Immutable
public class Constraint implements DbObject, TableNameAware {

    private final String tableName;
    private final String constraintName;
    private final ConstraintType constraintType;
    
    ...
}
```

Comment (mfvanek) [reply pg-index-health/src/main/java/io/github/mfvanek/pg/checks/host/CheckNotValidConstraintsOnHost.java:16]: For example see `IndexWithColumns` in https://github.com/mfvanek/pg-index-health/pull/377

**Comment by BLoHny** [comment]
> > The current test code doesn't seem to be good code. Could you help me?
> 
> 1. You can copy and modify existing class https://github.com/mfvanek/pg-index-health/blob/master/pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AddLinksBetweenAccountsAndClientsStatement.java in order to bring not valid FK.
> 2. You can add one more column to https://github.com/mfvanek/pg-index-health/blob/master/pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateTableWithCheckConstraintOnSerialPrimaryKey.java and not valid check constraint on it

```java
// AddInvalidForeignKeyStatement.java
    @Override
    public void execute(@Nonnull final Statement statement) throws SQLException {
        statement.execute(String.format("alter table if exists %1$s.accounts " +
                        "add constraint c_accounts_fk_invalid foreign key (invalid_client_id) references %1$s.invalid_clients (invalid_id);",
                schemaName));
    }
```

```java
// CreateTableWithCheckConstraintOnSerialPrimaryKey.java
    @Override
    public void execute(@Nonnull final Statement statement) throws SQLException {
        statement.execute(String.format("create table if not exists %1$s.another_table(" +
                        "id bigserial primary key, " +
                        "invalid_column integer, " +
                        "constraint not_reserved_id check (id > 1000), " +
                        "constraint less_than_million check (id < 1000000), " +
                        "constraint always_false check (invalid_column > 100 AND invalid_column < 0));",
                schemaName));
    }
```

Is this modified code suitable?

**Comment by mfvanek** [comment]
> Is this modified code suitable?

Unfortunately, no, it isn't.

When I say "not valid constraint" it's not about true or false constraint is, it's about whether database uses this constraint or not. Not valid constraint is like a draft. Please take a look at the original article https://habr.com/ru/articles/800121/

It should be like this
```
alter table if exists %1$s.accounts add constraint c_accounts_fk_client_id_not_validated_yet foreign key (client_id) references %1$s.clients (id) not valid;
```

In tests after you find all not valid constraints you can can call `validate constraint` and recheck the database: all issues should go away

**Comment by BLoHny** [comment]
```java
    @ParameterizedTest
    @ValueSource(strings = {PgContext.DEFAULT_SCHEMA_NAME, "custom"})
    void onDatabaseWithNotValidatedForeignKey(final String schemaName) {
        executeTestOnDatabase(schemaName, DatabasePopulator::withInvalidForeignKey, ctx ->
                assertThat(check)
                        .executing(ctx)
                        .hasSize(1)
                        .containsExactlyInAnyOrder(
                                Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_fk_client_id_not_validated_yet",
                                        ConstraintType.FOREIGN_KEY)
                        )
        );
    }
```
    
Is it okay to proceed with the test code like this?

**Comment by mfvanek** [comment]
Hi @BLoHny 

> Is it okay to proceed with the test code like this?

I'd suggest a few enhancements:

1. Rename `withInvalidForeignKey` to `withNotValidConstraints`
2. Add not valid FK and not valid check constraint
3. Add `dbp -> dbp.withNotValidConstraints().withUniqueConstraintOnSerialColumn()` to ensure that sql query ignores unique constraints and PK's

**Comment by BLoHny** [comment]
```sql
        statement.execute(String.format("alter table if exists %1$s.accounts " +
                        "add constraint c_accounts_chk_client_id_not_validated_yet check (client_id > 0) not valid;",
                schemaName));
```
After add more sql query..
```java
                        .containsExactly(
                                Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_chk_client_id_not_validated_yet",
                                        ConstraintType.CHECK),
                                Constraint.of(ctx.enrichWithSchema("accounts"), "c_accounts_fk_client_id_not_validated_yet",
                                        ConstraintType.FOREIGN_KEY)
                        )
```

add test code

proceed like this?

**Comment by mfvanek** [comment]
> proceed like this?

Yes, looks good

**Comment by BLoHny** [comment]
open new pull request https://github.com/mfvanek/pg-index-health/pull/382 by conflict

**Pull request closed without merging**